Tribler/community/tunnel/hidden_community.py

```diff
@@ -228,53 +258,103 @@ def monitor_downloads(self, dslist):
+
+    def on_dht_request(self, messages):
+        for message in messages:
+            info_hash = message.payload.info_hash
+
+            @call_on_reactor_thread
+            def dht_callback(info_hash, peers, _):
+                if not peers:
+                    peers = []
+                meta = self.get_meta_message(u'dht-response')
+                circuit_id = message.payload.circuit_id
+                # Send the list of peers for this info_hash back to the requester
+                dht_response_message = meta.impl(distribution=(self.global_time,), payload=(message.payload.circuit_id,
+                                                                                            message.payload.identifier,
+                                                                                            message.payload.info_hash,
+                                                                                            encode(peers)))
+                circuit = self.exit_sockets[circuit_id]
+                circuit.tunnel_data(message.candidate.sock_addr, TUNNEL_PREFIX + dht_response_message.packet)
+
+            self._logger.debug("Doing dht hidden seeders lookup for info_hash %s" % info_hash.encode('HEX'))
+            self.dht_lookup(info_hash, dht_callback)
+
+    def check_dht_response(self, messages):
+        for message in messages:
+            if not self.is_relay(message.payload.circuit_id):
+                request = self.request_cache.get(u"dht-request", message.payload.identifier)
+                if not request:
+                    yield DropMessage(message, "invalid dht-response identifier")
+                    continue
 
-    def do_lookup(self, info_hash):
-        # Get seeders from the DHT and establish an e2e encrypted tunnel to them
-        @call_on_reactor_thread
-        def dht_callback(info_hash, peers, _):
-            if not peers:
-                return
+            yield message
+
+    def on_dht_response(self, messages):
+        for message in messages:
+            self.request_cache.pop(u"dht-request", message.payload.identifier)
+
+            info_hash = message.payload.info_hash
+            _, peers = decode(message.payload.peers)
 
             blacklist = self.dht_blacklist[info_hash]
 
+            if self.notifier:
+                self.notifier.notify(NTFY_TUNNEL, NTFY_DHT_LOOKUP, info_hash.encode('hex')[:6], peers)
+
             # cleanup dht_blacklist
             for i in xrange(len(blacklist) - 1, -1, -1):
                 if time.time() - blacklist[i][0] > 60:
                     blacklist.pop(i)
-
             exclude = [rp[2] for rp in self.my_download_points.values()] + [sock_addr for _, sock_addr in blacklist]
-            for peer in set(peers):
+            for peer in peers:
                 if peer not in exclude:
+                    self.add_discovered_candidate(Candidate((peer[0], peer[1]), tunnel=False))
```

Comment:
I need the key of the introduction point to create a circuit that has the intropoint as required_endpoint (making it possible for the introduction point to respond on the same circuit), by adding the candidate manually it would introduce the candidate (like the walker does anyway with other candidates). 

Do you know another way to create a circuit between the downloader and intropoint which can be used by the intropoint to respond messages back?